dev/src/reference.ts

```diff
@@ -2982,99 +3048,62 @@ export class Query<
    * @return A Promise that will be resolved with the planner information, statistics
    *  from the query execution (if any), and the query results (if any).
    */
-  explain(
+  async explain(
     options?: firestore.ExplainOptions
   ): Promise<ExplainResults<QuerySnapshot<AppModelType, DbModelType>>> {
     if (options === undefined) {
       options = {};
     }
-
-    // Capture the error stack to preserve stack tracing across async calls.
-    const stack = Error().stack!;
-
-    return new Promise((resolve, reject) => {
-      let readTime: Timestamp;
-      let docs: Array<QueryDocumentSnapshot<AppModelType, DbModelType>> | null =
-        null;
-      let metrics: ExplainMetrics | null = null;
-
-      this._stream(undefined, options)
-        .on('error', err => {
-          reject(wrapError(err, stack));
-        })
-        .on('data', data => {
-          if (data.readTime) {
-            readTime = data.readTime;
-          }
-          if (data.document) {
-            if (docs === null) {
-              docs = [];
-            }
-            docs.push(data.document);
-          }
-          if (data.explainMetrics) {
-            metrics = data.explainMetrics;
-
-            if (docs === null && metrics?.executionStats !== null) {
-              // This indicates that the query was executed, but no documents
```

Comment:
This logic wasn't moved over. 

The effect is that returned snapshot will be null if there are no docs and `explainMetrics !== null` and `explainMetrics.executionStats === null`.

When `options.analyze !== true`, then only the `planSummary` is returned, but no execution actually happens. That means there will be no results, and no `executionStats`.

This is why the type `ExplainResults` has `readonly snapshot: T | null`

Reply:
Maybe I was mistaken, but this seemed like a just a long-winded way of achieving what I had instead which was just to initialise `docs` with an empty array to begin with.

I'm about to push a commit which changes the logic to only create a query snapshot if the stream received a `readTime` (and `docs` is still initialised to an empty array). So if the backend sends a readTime, when we get out a query snapshot with the received docs (or empty docs). If no read time received then no snapshot is created, and that should only happen with explain && analyze !== true (but is the prerogative of the backend rather than validating it in the lower levels of the SDK).

I then delegated further up the chain in the `_get()` function to verify that there is indeed a query snapshot present.

I also rationalised the `_get()`, `_getResponse()` and `_getResponseOrExplain()` methods:
- the internal `_get()` now returns response with metadata (and verifies that query snapshot exists). Previously this was just a trivial wrapper underneath the public `get()` of seemingly no use
- `_getReponse()` is the same but without the check for existence of a snapshot, and this is reused by `explain()`
- `_getResponseOrExplain()` removed
- Renamed `getResponse()` on `DocumentReader` to `_get()` to align with this convention

Reply:
SGTM